### PR TITLE
Feature/3498 cache connections

### PR DESCRIPTION
**Pull request by Darren-Osborne**

### CODE REVIEW

File: src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
+using System.Collections.Generic;
+using System.Net;
 using System.Net.Http.Json;
 using System.Text.Json;
 using System.Threading.Tasks;
 using FluentAssertions;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
 using RTGS.Bank.Simulator.Controllers.Handlers.IDCrypt.Connections;
+using RTGS.Bank.Simulator.Dtos;
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs:11]: Some unused usings in here

File: src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs [hunk 2]
```diff
@@ -90,4 +95,85 @@ await HttpResponse
...
+
+		[Fact]
+		public void WhenRequestSent_ThenOkResponse() =>
+			HttpResponse.StatusCode.Should().Be(HttpStatusCode.OK);
+
+		[Fact]
+		public async Task WhenRequestSent_ThenResponseMatchesExpected() => await Verifier.Verify(_connectionsResponse);
+	}
+
+	public class WithConnectionsInCacheButResetingCache : TestBase
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs:146]: Typo in the class name - "Resetting"

File: src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs [hunk 3]
```diff
@@ -90,4 +95,85 @@ await HttpResponse
...
+						CreatedAt = "2022-04-01"
+					}
+				}, TimeSpan.FromSeconds(50));
+
+			HttpResponse = await SimulatorHttpClient.GetAsync($"api/idCrypt/connections/true");
+		}
+
+		[Fact]
+		public void WhenRequestSent_ThenErrorResponse() =>
+			HttpResponse.StatusCode.Should().Be(HttpStatusCode.InternalServerError);
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs:177]: Why would we expect an internal server error when resetting the cache?

File: src/RTGS.Bank.Simulator.IntegrationTests/Extensions/IServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
 ﻿using IDCryptGlobal.Cloud.Agent.Identity;
+using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.IntegrationTests/Extensions/IServiceCollectionExtensions.cs:5]: Not added by you but there are unused usings here

File: src/RTGS.Bank.Simulator/Services/IdCryptConnectionsProvider.cs [hunk 5]
```diff
@@ -12,18 +13,45 @@ namespace RTGS.Bank.Simulator.Services;
...
 		_httpClient.BaseAddress = settings.Value.IdCryptApiAddress;
 		_httpClient.DefaultRequestHeaders.Add("X-API-Key", settings.Value.IdCryptApiKey);
 	}
 
 	public async Task<IList<ConnectionAccepted>> GetConnectionsAsync()
+	{
+		_logger.LogInformation("Getting connections from cache", ConnectionsPath);
+		async Task<IList<ConnectionAccepted>> Factory(ICacheEntry entry)
+		{
+			_logger.LogInformation("Connections not found in cache or cache expired.", ConnectionsPath);
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator/Services/IdCryptConnectionsProvider.cs:40]: `ConnectionsPath` is not used in the log message

File: src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+using Xunit;
+
+namespace RTGS.Bank.Simulator.Tests.Services;
+
+public class IdCryptConnectionsProviderTests
+{
+	private readonly IIdCryptConnectionsProvider _connectionsProvider;
+	private readonly FakeLogger<IdCryptConnectionsProvider> _fakeLogger = new();
+	private readonly Mock<HttpMessageHandler> _mockHandler = new();
+	public IdCryptConnectionsProviderTests()
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs:23]: Nitpick: missing an empty line

File: src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using Moq;
+using Moq.Protected;
+using RTGS.Bank.Simulator.Services;
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs:12]: Unused using

File: src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+	[Fact]
+	public async Task GetConnectionsAsync_WhenCacheIsResetCallsApiAgain()
+	{
+		await _connectionsProvider.GetConnectionsAsync();
+		_connectionsProvider.ResetCache();
+		await _connectionsProvider.GetConnectionsAsync();
+		_mockHandler.Protected().Verify("SendAsync", Times.Exactly(2), ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>());
+	}
+}
```
Comment (Bassetts) [inline src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs:74]: Should we have tests for the log messages? Other test classes seem to have them.

File: src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs (hunk 3 above)
Comment (Darren-Osborne) [reply src/RTGS.Bank.Simulator.IntegrationTests/Controllers/IdCrypt/GivenConnectionsRequest.cs:177]: Was just to prove that we werent getting data from the cache (i'd setup calls to the api to fail).
But - it did make it less than intuitative - changed to test that the data returned doesnt include the data from the cache

File: src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs (hunk 8 above)
Comment (Bassetts) [reply src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs:125]: Not sure if the other test classes do it, but I know we do in the SDK. Should we use strict ordering for the `BeEquivalentTo`? Happy to leave it if we don't elsewhere in here.

**Review by Bassetts — DISMISSED** [review]

File: src/RTGS.Bank.Simulator/ClientApp/src/pages/idCrypt/components/connection-list/styles.module.css [hunk 9]
```diff
@@ -9,6 +9,11 @@
 	background-color: background-2;
 }
 
+.container > Button {
```
Comment (sichiu) [inline src/RTGS.Bank.Simulator/ClientApp/src/pages/idCrypt/components/connection-list/styles.module.css:12]: I'd prefer to use the cs (classnames) node module to combine class names in your refreshButton function. I'm off tomorrow so perhaps Mike or Chris can help as they'll understand. There should be examples of this in the project.

File: src/RTGS.Bank.Simulator/Controllers/IdCryptController.cs [hunk 10]
```diff
@@ -44,10 +44,10 @@ public async Task<ActionResult> PollRefreshConnectionResponse(string alias, Canc
 
 
 	[HttpGet]
-	[Route("Connections")]
-	public async Task<ActionResult> Connections(CancellationToken cancellationToken)
+	[Route("Connections/{resetCache:bool?}")]
```
Comment (sichiu) [inline src/RTGS.Bank.Simulator/Controllers/IdCryptController.cs:47]: Should resetCache be a querystring?

File: src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs (hunk 8 above)
Comment (Darren-Osborne) [reply src/RTGS.Bank.Simulator.Tests/Services/IdCryptConnectionsProviderTests.cs:139]: We don't elsewhere in the simulator, but I concur we probably should. I've changed the code in the new class to do that now

**Review by Bassetts — APPROVED** [review]

**Pull request merged**